jobs/upstream.go

```diff
@@ -0,0 +1,108 @@
+// that are missing on the upstream.
+func SyncWithUpstream(ctx api.Context) error {
+	logger.Debugf("running upstream reconcile job")
+
+	jobHistory := models.NewJobHistory("SyncWithUpstream", api.UpstreamConf.Host, "")
+	_ = db.PersistJobHistory(ctx, jobHistory.Start())
+	defer func() {
+		_ = db.PersistJobHistory(ctx, jobHistory.End())
+	}()
+
+	reconciler := upstream.NewUpstreamReconciler(api.UpstreamConf, ReconcilePageSize)
+	for _, table := range api.TablesToReconcile {
+		if err := reconciler.SyncAfter(ctx, table, ReconcileMaxAge); err != nil {
+			jobHistory.AddError(err.Error())
+			logger.Errorf("failed to sync table %s: %v", table, err)
+		} else {
+			jobHistory.IncrSuccess()
+		}
+	}
+
+	return nil
+}
+
+// checkstatusSyncJob pushes new check statuses to upstream.
+type checkstatusSyncJob struct {
+	// created timestamp of the last check status that was pushed to upstream.
+	lastCreated time.Time
+}
+
+func (t *checkstatusSyncJob) Run() {
+	ctx, cancel := api.DefaultContext.WithTimeout(time.Minute)
+	defer cancel()
+
+	jobHistory := models.NewJobHistory("SyncCheckStatusesWithUpstream", api.UpstreamConf.Host, "")
+	_ = db.PersistJobHistory(ctx, jobHistory.Start())
+	defer func() {
+		_ = db.PersistJobHistory(ctx, jobHistory.End())
+	}()
+
+	if err := t.run(ctx); err != nil {
+		logger.Errorf("failed to run checkstatus sync job: %v", err)
+		jobHistory.AddError(err.Error())
+		return
+	}
+
+	jobHistory.IncrSuccess()
+}
+
+// SyncCheckStatusesWithUpstream pushes new check statuses to upstream.
+func (t *checkstatusSyncJob) run(ctx api.Context) error {
+	logger.Tracef("running checkstatus sync job since %s", t.lastCreated.Format(time.RFC3339))
+
+	var checkStatuses []models.CheckStatus
+	if t.lastCreated.IsZero() {
+		if err := ctx.DB().Select("check_statuses.*").
+			Joins("Left JOIN checks ON checks.id = check_statuses.check_id").
+			Where("checks.agent_id = ?", uuid.Nil).
+			Where("NOW() - check_statuses.created_at <= ?", ReconcileMaxAge).
+			Order("check_statuses.created_at").
+			Find(&checkStatuses).Error; err != nil {
+			return fmt.Errorf("failed to fetch checkstatuses: %w", err)
+		}
+	} else {
+		if err := ctx.DB().Select("check_statuses.*").
+			Joins("Left JOIN checks ON checks.id = check_statuses.check_id").
+			Where("checks.agent_id = ?", uuid.Nil).
+			Where("check_statuses.created_at > ?", t.lastCreated).
+			Order("check_statuses.created_at").
+			Find(&checkStatuses).Error; err != nil {
+			return fmt.Errorf("failed to fetch checkstatuses: %w", err)
+		}
+	}
+
+	if len(checkStatuses) == 0 {
+		return nil
+	}
+
+	logger.Tracef("Pushing %d check statuses to upstream", len(checkStatuses))
+	if err := upstream.Push(ctx, api.UpstreamConf, &upstream.PushData{AgentName: api.UpstreamConf.AgentName, CheckStatuses: checkStatuses}); err != nil {
+		return fmt.Errorf("failed to push check statuses to upstream: %w", err)
+	}
+
+	t.lastCreated = checkStatuses[len(checkStatuses)-1].CreatedAt
```

Comment:
I think we should still batch this with a limit of 500 while len(checkStatus) > 0